File: app/seed.py
from __future__ import annotations

import random
from datetime import date, timedelta
from calendar import monthrange

from sqlalchemy.orm import Session

from app.auth import hash_password
from app.models import (
    Department,
    Employee,
    Payroll,
    ProductService,
    Sale,
    Transaction,
    User,
)


DEMO_PHONE = "09131176583"
DEMO_PASSWORD = "123456789"


def _d(year: int, month: int, day: int) -> date:
    return date(year, month, day)


def seed_database(db: Session) -> None:
    if db.query(User).first():
        return

    rng = random.Random(42)
    today = date.today()

    user = User(
        phone=DEMO_PHONE,
        password_hash=hash_password(DEMO_PASSWORD),
        full_name="علی قضاوی",
    )
    db.add(user)

    departments = [
        Department(name="تولید"),
        Department(name="فروش و بازاریابی"),
        Department(name="مالی و اداری"),
        Department(name="فناوری اطلاعات"),
        Department(name="خدمات مشتریان"),
    ]
    db.add_all(departments)
    db.flush()

    roles = {
        "تولید": ["سرپرست خط", "اپراتور تولید", "کنترل کیفیت", "انباردار"],
        "فروش و بازاریابی": ["مدیر فروش", "کارشناس فروش", "کارشناس بازاریابی"],
        "مالی و اداری": ["حسابدار", "کارشناس منابع انسانی", "منشی"],
        "فناوری اطلاعات": ["توسعه‌دهنده", "پشتیبان سیستم", "تحلیل‌گر داده"],
        "خدمات مشتریان": ["کارشناس پشتیبانی", "هماهنگ‌کننده پروژه"],
    }
    first_names = ["رضا", "سارا", "محمد", "نرگس", "امیر", "زهرا", "حسین", "مینا", "پارسا", "الهام", "کیان", "نیلوفر"]
    last_names = ["محمدی", "احمدی", "کریمی", "رضایی", "حسینی", "موسوی", "جعفری", "نوری", "کاظمی", "اکبری"]

    employees: list[Employee] = []
    emp_idx = 1
    for dept in departments:
        count = 4 if dept.name != "تولید" else 7
        for _ in range(count):
            role = rng.choice(roles[dept.name])
            salary = {
                "سرپرست خط": 48_000_000,
                "اپراتور تولید": 32_000_000,
                "کنترل کیفیت": 36_000_000,
                "انباردار": 30_000_000,
                "مدیر فروش": 55_000_000,
                "کارشناس فروش": 38_000_000,
                "کارشناس بازاریابی": 35_000_000,
                "حسابدار": 40_000_000,
                "کارشناس منابع انسانی": 37_000_000,
                "منشی": 28_000_000,
                "توسعه‌دهنده": 52_000_000,
                "پشتیبان سیستم": 34_000_000,
                "تحلیل‌گر داده": 45_000_000,
                "کارشناس پشتیبانی": 31_000_000,
                "هماهنگ‌کننده پروژه": 39_000_000,
            }.get(role, 33_000_000)
            salary = int(salary * rng.uniform(0.92, 1.12))
            emp = Employee(
                code=f"EMP{emp_idx:03d}",
                full_name=f"{rng.choice(first_names)} {rng.choice(last_names)}",
                role=role,
                department_id=dept.id,
                monthly_salary=salary,
                hire_date=today - timedelta(days=rng.randint(90, 1600)),
                is_active=True,
            )
            employees.append(emp)
            emp_idx += 1
    db.add_all(employees)
    db.flush()

    catalog = [
        ("PRD-001", "بسته نرم‌افزار مالی سازمانی", "product", 85_000_000, 28_000_000),
        ("PRD-002", "ماژول گزارش‌گیری پیشرفته", "product", 42_000_000, 12_000_000),
        ("PRD-003", "لایسنس داشبورد هوش تجاری", "product", 28_000_000, 7_500_000),
        ("SRV-001", "پیاده‌سازی و استقرار سامانه", "service", 95_000_000, 45_000_000),
        ("SRV-002", "پشتیبانی ماهانه طلایی", "service", 18_000_000, 6_000_000),
        ("SRV-003", "آموزش تخصصی تیم مالی", "service", 24_000_000, 8_000_000),
        ("SRV-004", "مشاوره بهینه‌سازی هزینه", "service", 35_000_000, 10_000_000),
        ("PRD-004", "پکیج موبایل گزارش‌گیری", "product", 22_000_000, 6_500_000),
    ]
    items = [
        ProductService(
            code=code,
            name=name,
            kind=kind,
            unit_price=price,
            unit_cost=cost,
        )
        for code, name, kind, price, cost in catalog
    ]
    db.add_all(items)
    db.flush()

    # Sales + income/expense for last ~14 months
    start = today.replace(day=1) - timedelta(days=400)
    start = start.replace(day=1)
    cursor = start
    while cursor <= today:
        for item in items:
            # fewer weekends
            if cursor.weekday() >= 5 and rng.random() < 0.55:
                continue
            qty = rng.choice([0, 0, 0, 1, 1, 1, 2, 2, 3]) if item.kind == "product" else rng.choice([0, 0, 1, 1, 1, 2])
            if qty == 0:
                continue
            price = item.unit_price * rng.uniform(0.95, 1.08)
            cost = item.unit_cost * rng.uniform(0.97, 1.05)
            revenue = round(qty * price, 0)
            total_cost = round(qty * cost, 0)
            sale = Sale(
                item_id=item.id,
                sale_date=cursor,
                quantity=qty,
                unit_price=round(price, 0),
                unit_cost=round(cost, 0),
                revenue=revenue,
                cost=total_cost,
                profit=revenue - total_cost,
                channel=rng.choice(["فروش مستقیم", "وب‌سایت", "نماینده", "قرارداد سازمانی"]),
            )
            db.add(sale)
            db.add(
                Transaction(
                    txn_date=cursor,
                    direction="income",
                    category="فروش محصول" if item.kind == "product" else "ارائه خدمت",
                    title=f"فروش {item.name}",
                    amount=revenue,
                    source="sales",
                    reference=item.code,
                )
            )
            # بهای تمام‌شده کالا (هزینه کالای فروخته‌شده)
            db.add(
                Transaction(
                    txn_date=cursor,
                    direction="expense",
                    category="بهای تمام‌شده کالا",
                    title=f"هزینه کالای فروخته‌شده — {item.name}",
                    amount=total_cost,
                    source="cogs",
                    reference=item.code,
                )
            )

        # هزینه‌های عملیاتی
        if cursor.weekday() < 5 and rng.random() < 0.45:
            ops = [
                ("اجاره و تاسیسات", "پرداخت اجاره/انرژی", rng.randint(4_000_000, 12_000_000)),
                ("بازاریابی", "کمپین تبلیغاتی دیجیتال", rng.randint(2_000_000, 9_000_000)),
                ("لوازم مصرفی", "خرید ملزومات اداری", rng.randint(800_000, 3_500_000)),
                ("حمل‌ونقل", "هزینه ارسال و ماموریت", rng.randint(1_000_000, 4_500_000)),
                ("نگهداری تجهیزات", "سرویس سخت‌افزار", rng.randint(1_500_000, 6_000_000)),
            ]
            cat, title, amount = rng.choice(ops)
            db.add(
                Transaction(
                    txn_date=cursor,
                    direction="expense",
                    category=cat,
                    title=title,
                    amount=amount,
                    source="ops",
                )
            )
        cursor += timedelta(days=1)

    # Monthly payroll for last 14 months
    y, m = today.year, today.month
    for _ in range(14):
        for emp in employees:
            bonus = 0 if rng.random() < 0.7 else round(emp.monthly_salary * rng.uniform(0.05, 0.15), 0)
            deductions = round(emp.monthly_salary * rng.uniform(0.07, 0.11), 0)
            net = emp.monthly_salary + bonus - deductions
            paid_day = min(28, monthrange(y, m)[1])
            paid_on = _d(y, m, paid_day)
            if paid_on > today:
                continue
            db.add(
                Payroll(
                    employee_id=emp.id,
                    period_year=y,
                    period_month=m,
                    base_salary=emp.monthly_salary,
                    bonus=bonus,
                    deductions=deductions,
                    net_pay=net,
                    paid_on=paid_on,
                )
            )
            db.add(
                Transaction(
                    txn_date=paid_on,
                    direction="expense",
                    category="حقوق و دستمزد",
                    title=f"حقوق {emp.full_name} - {m:02d}/{y}",
                    amount=net,
                    source="payroll",
                    reference=emp.code,
                )
            )
        m -= 1
        if m == 0:
            m = 12
            y -= 1

    # Ensure today has some activity for daily report
    if not db.query(Sale).filter(Sale.sale_date == today).first():
        item = items[0]
        revenue = item.unit_price
        cost = item.unit_cost
        db.add(
            Sale(
                item_id=item.id,
                sale_date=today,
                quantity=1,
                unit_price=revenue,
                unit_cost=cost,
                revenue=revenue,
                cost=cost,
                profit=revenue - cost,
                channel="فروش مستقیم",
            )
        )
        db.add(
            Transaction(
                txn_date=today,
                direction="income",
                category="فروش محصول",
                title=f"فروش {item.name}",
                amount=revenue,
                source="sales",
                reference=item.code,
            )
        )

    db.commit()


def ensure_today_financial_data(db: Session) -> None:
    """If today has no transactions, generate sample sales/expenses for the daily dashboard."""
    today = date.today()
    has_today = (
        db.query(Transaction)
        .filter(Transaction.txn_date == today)
        .first()
        is not None
    )
    if has_today:
        return

    items = db.query(ProductService).filter(ProductService.is_active.is_(True)).all()
    if not items:
        return

    rng = random.Random(int(today.strftime("%Y%m%d")))
    for item in items:
        qty = rng.choice([1, 1, 2, 2, 3]) if item.kind == "product" else rng.choice([1, 1, 1, 2])
        price = item.unit_price * rng.uniform(0.96, 1.06)
        cost = item.unit_cost * rng.uniform(0.97, 1.04)
        revenue = round(qty * price, 0)
        total_cost = round(qty * cost, 0)
        db.add(
            Sale(
                item_id=item.id,
                sale_date=today,
                quantity=qty,
                unit_price=round(price, 0),
                unit_cost=round(cost, 0),
                revenue=revenue,
                cost=total_cost,
                profit=revenue - total_cost,
                channel="فروش مستقیم",
            )
        )
        db.add(
            Transaction(
                txn_date=today,
                direction="income",
                category="فروش محصول" if item.kind == "product" else "ارائه خدمت",
                title=f"فروش {item.name}",
                amount=revenue,
                source="sales",
                reference=item.code,
            )
        )
        db.add(
            Transaction(
                txn_date=today,
                direction="expense",
                category="بهای تمام‌شده کالا",
                title=f"هزینه کالای فروخته‌شده — {item.name}",
                amount=total_cost,
                source="cogs",
                reference=item.code,
            )
        )

    ops = [
        ("بازاریابی", "فعالیت بازاریابی روز جاری", rng.randint(2_500_000, 6_000_000)),
        ("لوازم مصرفی", "خرید ملزومات روز جاری", rng.randint(700_000, 2_200_000)),
        ("حمل‌ونقل", "هزینه ارسال روز جاری", rng.randint(1_000_000, 3_500_000)),
    ]
    cat, title, amount = rng.choice(ops)
    db.add(
        Transaction(
            txn_date=today,
            direction="expense",
            category=cat,
            title=title,
            amount=amount,
            source="ops",
        )
    )
    db.commit()


def migrate_legacy_labels(db: Session) -> None:
    """Rename legacy English/short categories to clear Persian labels."""
    renames = {
        "بهای تمام‌شده": "بهای تمام‌شده کالا",
        "COGS": "بهای تمام‌شده کالا",
        "OpEx": "هزینه‌های عملیاتی",
        "OPS": "هزینه‌های عملیاتی",
        "OPEX": "هزینه‌های عملیاتی",
    }
    changed = False
    for old, new in renames.items():
        rows = db.query(Transaction).filter(Transaction.category == old).all()
        for row in rows:
            row.category = new
            changed = True
    if changed:
        db.commit()

File: api/lib/seed.php
<?php
declare(strict_types=1);

function seed_demo_data(PDO $pdo): void
{
    $count = (int) $pdo->query('SELECT COUNT(*) FROM transactions')->fetchColumn();
    if ($count > 0) {
        return;
    }

    $deps = ['تولید', 'فروش و بازاریابی', 'مالی و اداری', 'فناوری اطلاعات', 'خدمات مشتریان'];
    $insDep = $pdo->prepare('INSERT INTO departments (name) VALUES (?)');
    foreach ($deps as $d) {
        $insDep->execute([$d]);
    }

    $depIds = $pdo->query('SELECT id, name FROM departments')->fetchAll();
    $depMap = [];
    foreach ($depIds as $row) {
        $depMap[$row['name']] = (int) $row['id'];
    }

    $employees = [
        ['EMP001', 'رضا محمدی', 'سرپرست خط', 'تولید', 48000000],
        ['EMP002', 'سارا احمدی', 'اپراتور تولید', 'تولید', 32000000],
        ['EMP003', 'محمد کریمی', 'کنترل کیفیت', 'تولید', 36000000],
        ['EMP004', 'نرگس رضایی', 'مدیر فروش', 'فروش و بازاریابی', 55000000],
        ['EMP005', 'امیر حسینی', 'کارشناس فروش', 'فروش و بازاریابی', 38000000],
        ['EMP006', 'زهرا موسوی', 'حسابدار', 'مالی و اداری', 40000000],
        ['EMP007', 'حسین جعفری', 'توسعه‌دهنده', 'فناوری اطلاعات', 52000000],
        ['EMP008', 'مینا نوری', 'کارشناس پشتیبانی', 'خدمات مشتریان', 31000000],
    ];
    $insEmp = $pdo->prepare(
        'INSERT INTO employees (code, full_name, role, department_id, monthly_salary, hire_date, is_active)
         VALUES (?, ?, ?, ?, ?, ?, 1)'
    );
    foreach ($employees as $e) {
        $insEmp->execute([$e[0], $e[1], $e[2], $depMap[$e[3]], $e[4], date('Y-m-d', strtotime('-400 days'))]);
    }

    $catalog = [
        ['PRD-001', 'بسته نرم‌افزار مالی سازمانی', 'product', 85000000, 28000000],
        ['PRD-002', 'ماژول گزارش‌گیری پیشرفته', 'product', 42000000, 12000000],
        ['PRD-003', 'لایسنس داشبورد هوش تجاری', 'product', 28000000, 7500000],
        ['SRV-001', 'پیاده‌سازی و استقرار سامانه', 'service', 95000000, 45000000],
        ['SRV-002', 'پشتیبانی ماهانه طلایی', 'service', 18000000, 6000000],
        ['SRV-003', 'آموزش تخصصی تیم مالی', 'service', 24000000, 8000000],
    ];
    $insItem = $pdo->prepare(
        'INSERT INTO products_services (code, name, kind, unit_price, unit_cost, is_active) VALUES (?, ?, ?, ?, ?, 1)'
    );
    foreach ($catalog as $c) {
        $insItem->execute($c);
    }
    $items = $pdo->query('SELECT * FROM products_services')->fetchAll();

    $insSale = $pdo->prepare(
        'INSERT INTO sales (item_id, sale_date, quantity, unit_price, unit_cost, revenue, cost, profit, channel)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)'
    );
    $insTx = $pdo->prepare(
        'INSERT INTO transactions (txn_date, direction, category, title, amount, source, reference)
         VALUES (?, ?, ?, ?, ?, ?, ?)'
    );
    $insPay = $pdo->prepare(
        'INSERT INTO payrolls (employee_id, period_year, period_month, base_salary, bonus, deductions, net_pay, paid_on)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?)'
    );

    $today = new DateTime('today', new DateTimeZone('Asia/Tehran'));
    $start = (clone $today)->modify('-120 days');
    $rng = 42;
    $pseudo = function () use (&$rng) {
        $rng = (1103515245 * $rng + 12345) & 0x7fffffff;
        return $rng / 0x7fffffff;
    };

    for ($d = clone $start; $d <= $today; $d->modify('+1 day')) {
        $ds = $d->format('Y-m-d');
        $dow = (int) $d->format('N');
        foreach ($items as $item) {
            if ($dow >= 6 && $pseudo() < 0.55) {
                continue;
            }
            $qty = $item['kind'] === 'product'
                ? [0, 0, 1, 1, 2][(int) floor($pseudo() * 5)]
                : [0, 0, 1, 1][(int) floor($pseudo() * 4)];
            if ($qty <= 0) {
                continue;
            }
            $price = (float) $item['unit_price'] * (0.95 + $pseudo() * 0.13);
            $cost = (float) $item['unit_cost'] * (0.97 + $pseudo() * 0.08);
            $revenue = round($qty * $price);
            $totalCost = round($qty * $cost);
            $insSale->execute([
                $item['id'], $ds, $qty, round($price), round($cost), $revenue, $totalCost, $revenue - $totalCost,
                'فروش مستقیم',
            ]);
            $cat = $item['kind'] === 'product' ? 'فروش محصول' : 'ارائه خدمت';
            $insTx->execute([$ds, 'income', $cat, 'فروش ' . $item['name'], $revenue, 'sales', $item['code']]);
            $insTx->execute([$ds, 'expense', 'بهای تمام‌شده کالا', 'هزینه ' . $item['name'], $totalCost, 'cogs', $item['code']]);
        }
        if ($dow < 6 && $pseudo() < 0.4) {
            $ops = [
                ['اجاره و تاسیسات', 8000000],
                ['بازاریابی', 4500000],
                ['لوازم مصرفی', 1500000],
                ['حمل‌ونقل', 2200000],
            ];
            $pick = $ops[(int) floor($pseudo() * count($ops))];
            $insTx->execute([$ds, 'expense', $pick[0], $pick[0], $pick[1], 'ops', null]);
        }
    }

    $emps = $pdo->query('SELECT id, code, full_name, monthly_salary FROM employees')->fetchAll();
    $y = (int) $today->format('Y');
    $m = (int) $today->format('n');
    for ($i = 0; $i < 6; $i++) {
        $yy = $y;
        $mm = $m - $i;
        while ($mm <= 0) {
            $mm += 12;
            $yy -= 1;
        }
        $paid = sprintf('%04d-%02d-28', $yy, $mm);
        if ($paid > $today->format('Y-m-d')) {
            continue;
        }
        foreach ($emps as $emp) {
            $bonus = $pseudo() < 0.7 ? 0 : round($emp['monthly_salary'] * 0.08);
            $ded = round($emp['monthly_salary'] * 0.09);
            $net = $emp['monthly_salary'] + $bonus - $ded;
            $insPay->execute([$emp['id'], $yy, $mm, $emp['monthly_salary'], $bonus, $ded, $net, $paid]);
            $insTx->execute([
                $paid, 'expense', 'حقوق و دستمزد',
                'حقوق ' . $emp['full_name'] . " - {$mm}/{$yy}",
                $net, 'payroll', $emp['code'],
            ]);
        }
    }
}
